### ISSUE TITLE
Popup Enhancements

### ISSUE THREAD

**Issue by philvessey** [arnaudleclerc/AzureMapsControl.Components#36]
Hi

Is it possible to enhance the popup support to bring in some of the extra options shown on https://docs.microsoft.com/en-us/azure/azure-maps/map-add-popup

<img width="234" alt="popup" src="https://user-images.githubusercontent.com/814882/112888343-22698180-90cc-11eb-8ff3-bdae67724980.png">

I have managed to create the above popup using the options already available but it would be great if we can get multiline support or the template support to use in the popups.

Enjoying the library so far - just starting out.

**Comment by philvessey** [comment]
Managed to achieve the multiline effect I needed using some line breaks.
Think the template options support would be good overall if that's something you can enhance.

<img width="225" alt="popup" src="https://user-images.githubusercontent.com/814882/112889579-bbe56300-90cd-11eb-9c8d-019934e77145.png">

**Comment by arnaudleclerc** [comment]
The template options is definitely something I need to look at. I need first to take care of adding the shapes to the events, so you can also rework the content of the popup according to the properties of a feature. It might take some time though, I currently don't have a lot of time to give to this project. I will let you know as soon as I make some progress

**Comment by philvessey** [comment]
Thanks - maybe I can get involved with some PRs to add some extra bits - this looks the best library I've seen for AzureMaps as a component so hopefully I can contribute.

**Comment by arnaudleclerc** [comment]
Sure, contributions are appreciated 🙂. I would advise you to wait until I have published a stable v1, there are a couple of differences between develop and the releases/1.0.0 branch. Once it's merged back into develop (eg. once I solved this serialisation issue with the shapes), feel free to take a look at it.

**Comment by philvessey** [comment]
Will do.

**Comment by arnaudleclerc** [comment]
FYI, I have published a v1.0.0, which means the develop branch is now up to date. I will write in the next couple of days some contributing guidelines and will take a look at incoming contributions from now on.

**Comment by arnaudleclerc** [comment]
FYI, I started the implementation of the popup template this afternoon. Here is a working example with a string template : 

```
@page "/Popups/StringTemplate"

@using AzureMapsControl.Components.Map
<AzureMap Id="map"
          EventActivationFlags="MapEventActivationFlags
                                .None()
                                .Enable(MapEventType.Ready)"
          OnReady="OnMapReady" />

@code  {

    public class PropertySubValues
    {
        public string SubValue { get; set; }
    }

    public async Task OnMapReady(MapEventArgs eventArgs)
    {
        var template = new Components.Popups.PopupTemplate
        {
            Content = new Components.Atlas.Either<string, IEnumerable<Components.Atlas.PropertyInfo>, IEnumerable<Components.Atlas.Either<string, IEnumerable<Components.Atlas.PropertyInfo>>>>
                ("This template uses a string template with placeholders.<br/><br/> - Value 1 = {value1}<br/> - Value 2 = {value2/subValue}<br/> - Array value [2] = {arrayValue/2}"),
            NumberFormat = new Components.Atlas.NumberFormatOptions
            {
                MaximumFractionDigits = 2
            }
        };

        var feature = new Components.Atlas.Feature<Components.Atlas.Point>(new Components.Atlas.Point(new Components.Atlas.Position(0, 0)), new Dictionary<string, object>
        {
            { "title", "Template 1 - String template" },
            { "value1", 1.2345678},
            { "value2", new PropertySubValues { SubValue = "Pizza" } },
            { "arrayValue", new [] { 3,4,5,6 } }
        });

        var popup = new Components.Popups.Popup(new Components.Popups.PopupOptions {
            Position = feature.Geometry.Coordinates,
            OpenOnAdd = true
        });

        await eventArgs.Map.AddPopupAsync(popup, template, feature.Properties);
    }

}
```

![image](https://user-images.githubusercontent.com/9578038/115768632-db10a100-a3aa-11eb-9dca-189b031d88d9.png)

There are still some missing fields on the `PopupTemplate` class and I would like to test more difficult templates before publishing it, but this looks promising.

**Comment by philvessey** [comment]
Looks a really good start!

**Comment by arnaudleclerc** [comment]
I just published a version `1.5.0-alpha0016` adding the support of the popup templates. If you take a look at the samples on the `develop` branch, you will find the same examples as on https://docs.microsoft.com/en-us/azure/azure-maps/map-add-popup under the `Popups` folder. 

In short, there is two ways to use the template : 

- Use the `AddPopupAsync` method of the `Map` and pass an instance of `Popup`, `PopupTemplate` and the properties.
- Use the `ApplyTemplateAsync` method of a `Popup` which has already been added to the map and pass the template, the properties and optionally an update of the options.

String templates, `PropertyInfo` templates and the combination of both are supported : 

```
@page "/Popups/MultipleContentTemplate"

@using AzureMapsControl.Components.Map
<AzureMap Id="map"
          EventActivationFlags="MapEventActivationFlags
                                .None()
                                .Enable(MapEventType.Ready)"
          OnReady="OnMapReady" />

@code  {

    public class PropertySubValues
    {
        public string SubValue { get; set; }
    }

    public async Task OnMapReady(MapEventArgs eventArgs)
    {
        var templateContent = new[] {
        new Components.Atlas.Either<string, IEnumerable<Components.Atlas.PropertyInfo>>
        ("This template has two pieces of content; a string template with placeholders and a array of property info which renders a full width image.<br/><br/> - Value 1 = {value1}<br/> - Value 2 = {value2/subValue}<br/> - Array value [2] = {arrayValue/2}"),
        new Components.Atlas.Either<string, IEnumerable<Components.Atlas.PropertyInfo>>
        (
            new [] {
                new Components.Atlas.PropertyInfo {
                    PropertyPath = "imageLink",
                    Label = "Image",
                    HideLabel = true,
                    HyperlinkFormat = new Components.Atlas.FormatOptions.HyperLinkFormatOptions {
                        IsImage = true
                    }
                }
            }
        )};

        var template = new Components.Popups.PopupTemplate
        {
            Content = new Components.Atlas.Either<string, IEnumerable<Components.Atlas.PropertyInfo>, IEnumerable<Components.Atlas.Either<string, IEnumerable<Components.Atlas.PropertyInfo>>>>
                (templateContent),
            NumberFormat = new Components.Atlas.FormatOptions.NumberFormatOptions
            {
                MaximumFractionDigits = 2
            }
        };

        var properties = new Dictionary<string, object>
        {
            { "title", "Template 3 - Multiple content template" },
            { "value1", 1.2345678},
            { "value2", new PropertySubValues { SubValue = "Pizza" } },
            { "arrayValue", new [] { 3,4,5,6 } },
            { "imageLink", "https://azuremapscodesamples.azurewebsites.net/common/images/Pike_Market.jpg"}
        };

        var position = new Components.Atlas.Position(0, 0);

        var popup = new Components.Popups.Popup(new Components.Popups.PopupOptions {
            Position = position,
            OpenOnAdd = true
        });

        await eventArgs.Map.AddPopupAsync(popup, template, properties);
    }

}
```

I will try to publish a stable version this weekend. Feel free to try this version and let me know if you have any feedback!

**Comment by arnaudleclerc** [comment]
I just published a v 1.5.0 with those changes. I will close this issue, but feel free to reopen it if necessary.
